backend/test/resolve-payload-cids.test.js

```
@@ -212,4 +212,92 @@ describe('deal-observer-backend piece indexer payload retrieval', () => {
     await resolvePayloadCids(fetchMinerId, resolvePayloadCid, pgPool, 10000, now)
     assert.strictEqual(payloadsCalled, 1)
   })
+
+  it('piece indexer count number of resolved paylod CIDs', async () => {
+    const defaultActieDeal = Value.Parse(ActiveDealDbEntry, {
+      miner_id: 1,
+      piece_cid: 'baga1',
+      client_id: 1,
+      activated_at_epoch: 1,
+      piece_size: 1000,
+      term_start_epoch: 1,
+      term_min: 1,
+      term_max: 1,
+      sector_id: 1,
+      payload_cid: undefined,
+      payload_retrievability_state: PayloadRetrievabilityState.NotQueried,
+      last_payload_retrieval_attempt: undefined
+    })
+    await storeActiveDeals([defaultActieDeal], pgPool)
+    await storeActiveDeals([{ ...defaultActieDeal, miner_id: 2, payload_cid: 'PAYLOAD_CID', payload_retrievability_state: PayloadRetrievabilityState.Resolved }], pgPool)
+    await storeActiveDeals([{ ...defaultActieDeal, miner_id: 3, payload_cid: 'PAYLOAD_CID', payload_retrievability_state: PayloadRetrievabilityState.Resolved }], pgPool)
+    const missingPayloadCids = await countStoredActiveDealsWithPayloadState(pgPool, PayloadRetrievabilityState.Resolved)
+    assert.strictEqual(missingPayloadCids, 2n)
+  })
+
+  it('piece indexer count number of unresolved paylod CIDs', async () => {
+    const defaultActieDeal = Value.Parse(ActiveDealDbEntry, {
+      miner_id: 1,
+      piece_cid: 'baga1',
+      client_id: 1,
+      activated_at_epoch: 1,
+      piece_size: 1000,
+      term_start_epoch: 1,
+      term_min: 1,
+      term_max: 1,
+      sector_id: 1,
+      payload_cid: undefined,
+      payload_retrievability_state: PayloadRetrievabilityState.NotQueried,
+      last_payload_retrieval_attempt: undefined
+    })
+    await storeActiveDeals([defaultActieDeal], pgPool)
+    await storeActiveDeals([{ ...defaultActieDeal, miner_id: 2, payload_retrievability_state: PayloadRetrievabilityState.Unresolved }], pgPool)
```

What happens when we need to add more deals to the test setup here and we forget that each deal must be unique - e.g. we forget to change `miner_id`.

```js
await storeActiveDeals([{ ...defaultActieDeal, miner_id: 2, payload_retrievability_state: PayloadRetrievabilityState.Unresolved }], pgPool)
await storeActiveDeals([{ ...defaultActieDeal, miner_id: 2, payload_retrievability_state: PayloadRetrievabilityState.TerminallyUnretrievable }], pgPool)
```

Will the new version fail and tell us about the violation of the unique constraint?

Or will it silently ignore the duplicate deal, leading us to believe the code works as intended, which is not true?

---

I believe the fix intended was that it would ignore duplicate deals, no?

---

I believe the intent of the hot fix that we did last week was that the unique constraint is ignored on duplicate values, no?

---

We discussed this thread yesterday on a Zoom call, it's related to https://github.com/CheckerNetwork/spark-deal-observer/pull/114#discussion_r1958483960

I find it important to change the tests to use a different method for creating test-data deal than `storeActiveDeals`. When inserting test-data deals to the DB, we don't want to ignore conflicts (violations of the unique constraint), we want tests to fail right at the point where the attempt to insert a duplicate deal.

---

Addressed in de7ad74